pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -56,70 +63,132 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
         }
 
+    /** Builds an [ASTMethod] wrapper for the [MethodDeclaration] node. */
+    private fun buildMethodDeclaration(node: MethodDeclaration, parent: ApexNode<*>?) =
+        when {
+            node.isAnonymousInitializationCode() && !node.hasKeyword(Keyword.STATIC) ->
+                build(node.body, parent)
+            else -> {
+                // Rename the method if it's a static initializer or constructor to internal name
+                val methodDecl =
+                    when {
+                        node.isAnonymousInitializationCode() ->
+                            cloneMethod(node, newName = ASTMethod.STATIC_INIT_ID)
+                        node.isConstructor -> cloneMethod(node, newName = ASTMethod.CONSTRUCTOR_ID)
+                        else -> node
+                    }
+                ASTMethod(methodDecl).apply {
+                    buildModifiers(node.modifiers).also { it.setParent(this) }
+                    buildChildren(methodDecl, parent = this, exclude = { it in methodDecl.modifiers })
+                }
+            }
+        }
+
+    /** Builds an [ASTProperty] wrapper for the [PropertyDeclaration] node. */
+    private fun buildPropertyDeclaration(node: PropertyDeclaration) =
+        ASTProperty(node).apply {
+            buildModifiers(node.modifiers).also { it.setParent(this) }
+
+            // Build accessors
+            val internalName = ASTProperty.formatAccessorName(this)
+            listOfNotNull(node.getter, node.setter).forEach { accessor ->
+                // Rename the accessor to internal name
+                val newAccessor = cloneMethod(accessor, newName = internalName)
+                buildAndSetParent(newAccessor, parent = this)
+            }
+
+            buildChildren(
+                node,
+                parent = this,
+                exclude = { it in node.modifiers || it == node.getter || it == node.setter }
+            )
+        }
+
+    /** Builds an [ASTFieldDeclarationStatements] wrapper for the [FieldDeclarationGroup] node. */
+    private fun buildFieldDeclarationGroup(node: FieldDeclarationGroup) =
+        ASTFieldDeclarationStatements(node).apply {
+            buildModifiers(node.modifiers).also { it.setParent(this) }
+            buildChildren(node, parent = this, exclude = { it in node.modifiers })
+        }
+
     /** Builds an [ASTModifierNode] wrapper for the list of [Modifier]s. */
     private fun buildModifiers(modifiers: List<Modifier>) =
         ASTModifierNode(ListAdapter(modifiers, loc = SourceLocation.UNKNOWN)).apply {
-            build(modifiers, parent = this)
+            modifiers.forEach { buildAndSetParent(it, parent = this) }
         }
 
+    /**
+     * Creates a new [MethodDeclaration] with the same properties as [src] but with a [new name]
+     * [newName].
+     */
+    private fun cloneMethod(src: MethodDeclaration, newName: String) =
+        MethodDeclaration(
```

Comment:
It doesn't seem right that the (PMD) tree builder is creating Summit nodes.

How inconvenient would it be to handle these as special cases inside the pmd.ASTMethod?  Or, if that is a bad solution, what would it look like to push these changes upstream into the original Summit tree?

Reply:
The motivation behind this was to match the `image` of `ASTMethod` to the one expected by rules (which is a hard-coded value based on Jorje). I notice that instead of changing the name directly on the `MethodDeclaration`, we can also modify the `image` getter for the same result:
```java
@Override
public String getImage() {
    if (node.isAnonymousInitializationCode()) {
        return STATIC_INIT_ID;
    } else if (node.isConstructor) {
        return CONSTRUCTOR_ID;
    } else {
        return node.getId().asCodeString();
    }
}
```
How does this sound?

Reply:
Sounds great! Thanks for exploring the alternate approach.